feat extra MEH and HYPER combined modifier and copilot key support

nicenemo opened the pull request:
Add left and right combinations of modifiers such as HYPER and MEH.
Adds the copilot key.
Closes #2339 

Tested on my Cornish-Zen with https://github.com/nicenemo/zmk-config/tree/test_custom_modifier_patch_for_zmk_config

caksoylar commented:
Note that I haven't given this as detailed a look as lesshonor yet, but I have some high level thoughts:
- I don't believe that having all combinations pre-defined is particularly useful. e.g. `LS(LA(key))` is not less unwieldy than remembering `LSAK(key)`, and the former has the advantage of implying/teaching to the user that the functions are easily combined in an intuitive way
  - I can sort of see `HYPER` and `MEH` as special cases since they are used in the wider keyboard community and it isn't immediately obvious what modifiers they are composed of
- I strongly agree with @lesshonor on the point of namings breaking existing conventions w.r.t. `*K` pattern, as well as the modifier side being on the end rather than the beginning
- I also agree about the comment on us not needing an extra docs page for this

nicenemo commented on app/include/dt-bindings/zmk/extra_modifiers.h:
That is exactly the kind of review comment I was hoping for. Thank you! I tried several options in my own repo first but was not sure yet.
I did keep them separate for now to not cause any merge conflicts with other stuff. If more people agree to merge them with the rest, I will do that.

nicenemo commented:
The hyper I use on a daily basis on Windows.
MEH is useful in combination with something like autohotkey since it is as akward as hyper and not many programs used it for hotkeys. This means you can define your own.

The combinations with  Control shift or alt shift are useful in for example powershell since you cannot define hotkeys with the Windows key in Powershell. 
Also handy if you want to define a copilot key ;)
MEH and Hyper are available in QMK, but only in a left variant.

So What should I do to make this more acceptable?
1. merge the documentation and definitions in existing .h and doc files?
2.  drop the 2 key combinations? Or do other peopele also think this is useful

nicenemo commented:
If I:
1. Remove the 2 key modifiers 
2. move meh and hyper to existing include with proper renaming 
3. move and adjust the documentation
4. add office aliases for hyper
5. add a copilot alias


Would that be acceptable?

nicenemo commented:
I did the following:
* Renamed the Hyper and MEH keys. For MEH i followd QMK's solution, e.g. KC_LMEH and `LMEH(...)`
* I removed all my two key combined modifiers
* Updated the documentation
* moved the extra_modifiers into existing files for definitions and modifiers
* Added a Copilot key
* Fixed documentation build errors I made

Any further suggestions for improvement?
Should I provide a squashed pull request and close this one?

nicenemo commented:
> You don't need to close this PR and open a new one to squash commits—rebasing and force pushing is fine. [Adhering to conventional commit guidelines is encouraged](https://github.com/zmkfirmware/zmk/blob/main/CONTRIBUTING.md#commit-messages). (Don't worry about preserving any of the co-author stuff; I don't care.)
> 
> ZMK is not an ultra-fast moving project; there are _maybe_ three people with merge permissions (this does not include me). Your patience with discussions and reviews is appreciated. Taking the extra time to carefully audit your proposed changes for any typos and inconsistencies would be productive.

I just want it to be valuable to many people. The best possible naming is important here. Changing it later is painful. I like your naming solutions for the functions. 
I will accept your suggestions.

I do understand and respect that some do not see the value of defines for 2 key modifiers. 

Regarding the concrete example, I do agree too. I initially did not to keep it similar with the rest.

Regarding the missing back tick, locally I did not get an error. 

How do I squash the commits on the same branch? Not being hit fluent, sorry.

nicenemo commented:
Support on iOS, MacOS and Android was untested yet.  Claiming it to be supported is a bit too bold and preliminary to my taste. 

I will test again  on Windows, Linux. I will test Android and iOS. I do not have an Mac available. Can anybody test that?

The Copilot Key only can work as intended on Windows 11 of course.
On Linux The Hyper and MEH are supported. Higher function Keys such as F23 combined with modifiers or not sometimes give suprising results depending on Desktop environment or even lower level settings in WayLand, X or below.

What does supported exactly mean? I will ask for clarification on discord.
How about iPad OS and ChromeOS support? There is no room for that in the documentation yet.

nicenemo commented:
I tested on Windows, Linux(KDE) and  i tried iOS. I do not have a Mac so I cannot test that.
The iOS tests are inconclusive because of some iOS device issues and I am less familiar with iOS.
Hyper and MEH work fine on Windows and Linux. The Copilot key starts a search on some configurations of Windows 11 and will probably start copilot if configured. On Windows 10 it will select the first icon on the Desktop. On Linux  higher function key support varies sice. It depends on your Desktop environment configuration, X, Wayland or something  more low level. Therefore I did not set that to true or false.
Given the previous discussions, and changes I made based on the input , I think this pull request is ready to be reviewed.
